### ISSUE TITLE
DHF 4.0.0: mlDeploy fails (in some conditions) if project contains REST extension in ml-config

### ISSUE THREAD

**Issue by kosed** [marklogic/marklogic-data-hub#1378]
#### The issue

Running mlDeploy for the first time will fail when there is a REST extension in src/main/ml-modules/services.
The task actually fails while trying to load the modules.
mlDeploy will keep failing until ml(re)LoadModules is run.

My guess is that this is due to the staging server using a custom rewriter.
The rewriter config needs to be loaded in the modules DB first, before serving requests (like creating a new REST extensions).
This also explains why it only happens before a successful modules loading.

#### Tech details

Which Operating System are you using?
MacOS

Which version of MarkLogic are you using?
9.0-6.2

Which version of the Data Hub Framework are you using?
4.0.0

Is this a QuickStart UI bug? If so, which browser are you using?
No

#### The devil is in the details...

```
End output of: ./gradlew -i mlDeploy
```
```
Executing command [com.marklogic.appdeployer.command.schemas.LoadSchemasCommand] with sort order [350]
Loading schemas from path: /ml/src/customers/GAIG/bdhf400/src/main/hub-internal-config/schemas
Initializing ExecutorService
Finished loading schemas from: /ml/src/customers/GAIG/bdhf400/src/main/hub-internal-config/schemas
Releasing connection
Finished executing command [com.marklogic.appdeployer.command.schemas.LoadSchemasCommand]

Executing command [com.marklogic.appdeployer.command.clusters.ModifyLocalClusterCommand] with sort order [350]
No resource directory found at: /ml/src/customers/GAIG/bdhf400/src/main/hub-internal-config/clusters
Finished executing command [com.marklogic.appdeployer.command.clusters.ModifyLocalClusterCommand]

Executing command [com.marklogic.appdeployer.command.modules.LoadModulesCommand] with sort order [400]
Initializing new instance of ModulesLoader
Loading asset modules from dir: src/main/ml-modules
Initializing ExecutorService
Shutting down ExecutorService
Loading all non-asset modules from dir: src/main/ml-modules
Initializing ExecutorService
Loading dummy-ext resource extension from file dummy-ext.xqy

Writing resource services source for dummy-ext
Error occurred while loading modules; host: localhost; port: 20015; cause: Local message: config/resources not found for write. Server Message: Server (not a REST instance?) did not respond with an expected REST Error message.
com.marklogic.client.ResourceNotFoundException: Local message: config/resources not found for write. Server Message: Server (not a REST instance?) did not respond with an expected REST Error message.
        at com.marklogic.client.impl.OkHttpServices.putPostValueImpl(OkHttpServices.java:2872)
        at com.marklogic.client.impl.OkHttpServices.putValue(OkHttpServices.java:2751)
        at com.marklogic.client.impl.ResourceExtensionsImpl.writeServices(ResourceExtensionsImpl.java:206)
        at com.marklogic.client.ext.modulesloader.impl.DefaultModulesLoader.lambda$installService$1(DefaultModulesLoader.java:479)
        at com.marklogic.client.ext.modulesloader.impl.DefaultModulesLoader.lambda$executeTask$5(DefaultModulesLoader.java:570)

> Task :mlDeployApp FAILED
Shutting down ExecutorService
Releasing connection
:mlDeployApp (Thread[Task worker for ':',5,main]) completed. Took 1 mins 0.085 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':mlDeployApp'.
> Error occurred while loading REST modules: Error occurred while loading modules; host: localhost; port: 20015; cause: Local message: config/resources not found for write. Server Message: Server (not a REST instance?) did not respond with an expected REST Error message.
```

**Comment by damonfeldman** [comment]
this was also typed up with different wording. Capturing here for completeness:

An issue with DHF4 and smart-mastering: which is also an issue with marklogic-unit-test. In DHF4, the staging app server has a custom rewriter - /data-hub/4/rest-api/rewriter.xml . That module is loaded via hubInstallModules. But the modules associated with mlRestApi dependencies are loaded before hub modules. So on your first deployment, SM or MUT will fail to deploy because calls are made to /v1/config/* on the staging REST server, but it can't find its REST rewriter, and thus you get 404s back. Possible workarounds would ensure the hub modules to be loaded before the mlRestApi modules are loaded.

**Comment by grechaw** [comment]
Argh, so frustrating.  There's quite a deployment dance with the appservers because of where the rewriter sits.  The exact order of operations will be important to track down.    Thanks for the repro as we've not been able to see when this issue surfaces.

**Comment by grechaw** [comment]
https://project.marklogic.com/jira/browse/DHFPROD-1421

**Comment by damonfeldman** [comment]
Possible workaround from Edgard:

> remove the modules deployment commands from mldeploy, then run mlreloadmodules after mlDeploy
> 
> ext {
>     mlAppDeployer.removeCommand("LoadHubModulesCommand")
>     mlAppDeployer.removeCommand("LoadModulesCommand")
>     mlAppDeployer.removeCommand("LoadUserStagingModulesCommand")
> }

**Comment by kkanthet** [comment]
Fixed as part of Release 4.1.0
